codespace/liveshare keepalive: keep LS sessions alive with PF traffic

josebalius opened the pull request:
This PR introduces new behavior to the liveshare package and codespace ssh command that enables it to notify the codespace server host of SSH activity and keep alive the Live Share session.

You'll notice that other types of port forwarding do not keep alive and this is by design, only the SSH activity can keep the session alive, user app port traffic is not considered enough activity to keep alive.

I have tested this against PPE codespaces and it works quite well. There are further changes necessary server-side to make better use of this activity signal but that will happen separate of this.

- `liveshare.Session` now implements a `heartbeat` that ticks on an interval, checks if there is a keep alive reason and notifies the server host of it. The one-minute duration is what the platform folks recommend.
- `liveshare.PortForwarder` now accepts a `keepAlive` parameter, which when true will make the PF monitor traffic and notify the session of the traffic type (`input` or `output`) as it happens. This notification is non-blocking and non-critical, meaning if multiple events happen in the span of the keep alive interval, we only retain the first and drop the others.
- The SSH command introduces a new flag `--debug` which when set, outputs the path of a temporary file where debugging logs are stored. For right now, the only thing that logs there is the heartbeat, we could add more in the future.
- The above requirement makes us introduce a `fileLogger` into the code to log into the file. I really wanted to us `log.Logger` here and have it as a TODO.
- Updates all call sites that user the port forwarder to deal with new params.
- Tests

josebalius committed:
Live Share session activity detection

- Session now accepts two new options: ClientName and Logger
- Port forwarder now supports a keepAlive parameter which when true,
  instructs the PF to call the session's keepAlive method.
- Port forwarder uses a new traffic monitor to detect I/O traffic and
  notify the session when applicable.
- The SSH command introduces a new debug flag which enables the command
  to log to a new temporary file. The file path is printed to the user.

josebalius committed:
Merge branch 'trunk' of github.com:cli/cli into jg/liveshare-keepalive

josebalius committed:
Tests for most of the new behavior

- Made the heartbeat interval configurable for easier testing
- Moved span to the top of connect to capture the full execution

josebalius committed:
Make fileLogger more versatile

josebalius committed:
Fix races and remove unreachable code

josebalius committed:
Fix unrelated tests

josebalius committed:
Fix additional race in tests

adonovan commented on pkg/cmd/codespace/ssh.go:
```diff
@@ -103,3 +124,35 @@ func (a *App) SSH(ctx context.Context, sshArgs []string, sshProfile, codespaceNa
+}
+
+// newFileLogger creates a new fileLogger. It returns an error if the file
+// cannot be created. The file is created in the operating system tmp directory
```
/tmp is a slightly odd place to write logs. Does the gh tool perhaps have a precedent we can follow?

josebalius replied:
Yeah, I have thoughts here:
1. See if the CLI has a precedent and follow that
2. Perhaps make `debug` a string flag that accepts a path to write the logs to and if empty then write to /tmp.
3. Just write to the same place the command is being run from.
4. Keep as is and write to tmp.

josebalius replied:
I spoke to @mislav on this. There is no real precedent in the CLI for this but we decided to go with option (2) here and allow the user to specify the file if they wish to do and if not default to a temp file.

adonovan commented on pkg/liveshare/port_forwarder.go:
```diff
@@ -133,8 +157,21 @@ func (fwd *PortForwarder) handleConnection(ctx context.Context, id channelID, co
+		connReader    io.Reader = conn
+	)
+
+	// If we the port forwader has been configured to keep the session alive
```
"If the"
"forwarder"

adonovan commented on pkg/liveshare/session.go:
```diff
@@ -97,3 +102,42 @@ func (s *Session) StartSSHServer(ctx context.Context) (int, string, error) {
 	return port, response.User, nil
 }
+
+// heartbeat ticks every interval and sends a signal to the Live Share host to keep
```
// heartbeat runs until context cancellation, periodically checking whether there is a
// reason to keep the connection alive, and if so, notifying the Live Share host to do so.

s/interval/period/ to emphasize that it is recurring.

josebalius replied:
Can you do a class on docs writing? Always so good 😄

adonovan commented on pkg/liveshare/session_test.go:
```diff
@@ -221,3 +227,180 @@ func TestInvalidHostKey(t *testing.T) {
+
+	// if keepAlive blocks, we'll never reach this and timeout the test
+	// timing out
+	if i != 2 {
```
This condition is provably true.

adonovan commented on pkg/liveshare/session_test.go:
```diff
@@ -221,3 +227,180 @@ func TestInvalidHostKey(t *testing.T) {
+	// if keepAlive blocks, we'll never reach this and timeout the test
+	// timing out
+	if i != 2 {
+		t.Errorf("unexpected iteration account, expected: 2, got: %d", i)
```
"count"

adonovan commented:
LGTM, but I can't approve for cli/cli.

josebalius commented:
>LGTM, but I can't approve for cli/cli.

Hmm I thought @cli/codespaces could approve? perhaps not.

josebalius committed:
Update docs, remove needless condition check

josebalius committed:
Add more time between events

adonovan reviewed (approved):
Ah, I was confused by the lack of my name among the reviewers at the bottom of the Conversation pane. But I see now that Code > Review changes presents an Approve button. Approved!

Sorry for the confusion!

josebalius committed:
Switch to standard lib log.Logger & support dfile

- --debug-file flag can now be used in conjuction with --debug to
  specify the debug file path
- Push out logger concerns to callers of liveshare

josebalius commented on pkg/liveshare/port_forwarder.go:
```diff
@@ -15,16 +15,19 @@ type PortForwarder struct {
-func NewPortForwarder(session *Session, name string, remotePort int) *PortForwarder {
+// of the remote port or service. The keepAlive flag indicates whether
+// the session should be kept alive with port forwarding traffic.
+func NewPortForwarder(session *Session, name string, remotePort int, keepAlive bool) *PortForwarder {
```
@mislav I know we chatted on this, but after more thinking some more I prefer to keep this API instead. Using a `pf.SetKeepAlive` setter just doesn't feel right, and tbh using setters/getters in Go feels like an anti-pattern. There are several opinions on this of course but just FYI as to why the diff here did not change.

josebalius committed:
Merge branch 'trunk' of github.com:cli/cli into jg/liveshare-keepalive

mislav reviewed (approved):
Thanks for the hard work! The keepalive solutions are clever and logging looks good. There is some more room to clean up various "logger" objects within `codespace` commands but we can do as polish next steps.
